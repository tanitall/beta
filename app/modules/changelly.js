"use strict";
module.exports = (function() {
  let URL = "https://api.changelly.com";
  let io = require("socket.io-client");
  let jayson = require("jayson");
  let crypto = require("crypto");
  let client = jayson.client.https(URL);

  function Changelly(apiKey, apiSecret) {
    this._id = function() {
      return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
        c
      ) {
        let r = (Math.random() * 16) | 0,
          v = c == "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      });
    };

    this._sign = function(message) {
      return crypto
        .createHmac("sha512", apiSecret)
        .update(JSON.stringify(message))
        .digest("hex");
    };

    this._request = function(method, options, callback) {
      let id = this._id();
      let message = jayson.utils.request(method, options, id);
      client.options.headers = {
        "api-key": apiKey,
        sign: this._sign(message)
      };

      client.request(method, options, id, function(err, response) {
        callback(err, response);
      });
    };

    let self = this;

    this._socket = io.connect(URL, {
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      reconnectionAttempts: "Infinity"
    });

    this._socket.on("connect", function() {
      let message = {
        Login: {}
      };

      self._socket.emit("subscribe", {
        apiKey,
        sign: self._sign(message),
        message
      });
    });
  }

  Changelly.prototype = {
    getCurrencies(callback) {
      return this._request("getCurrencies", {}, callback);
    },
    createTransaction(from, to, address, amount, extraId, callback) {
      let params = {
        from,
        to,
        address,
        amount,
        extraId
      };

      return this._request("createTransaction", params, callback);
    },
    getMinAmount(from, to, callback) {
      let params = {
        from,
        to
      };

      return this._request("getMinAmount", params, callback);
    },
    getExchangeAmount(from, to, amount, callback) {
      let params = {
        from,
        to,
        amount
      };

      return this._request("getExchangeAmount", params, callback);
    },
    getTransactions(limit, offset, currency, address, extraId, callback) {
      let params = {
        limit,
        offset,
        currency,
        address,
        extraId
      };

      return this._request("getTransactions", params, callback);
    },
    getStatus(id, callback) {
      let params = {
        id
      };

      return this._request("getStatus", params, callback);
    },
    on(channel, callback) {
      this._socket.on(channel, callback);
    }
  };

  return Changelly;
})();
