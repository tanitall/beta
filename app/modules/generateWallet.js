import {
  getAccountsFromWIFKey,
  generatePrivateKey,
  getWIFFromPrivateKey,
  encrypt_wif
} from "neon-js";

// Constants
export const NEW_WALLET_KEYS = "NEW_WALLET_KEYS";
export const NEW_WALLET = "NEW_WALLET";
export const SET_GENERATING = "SET_GENERATING";
export const RESET_KEY = "RESET_KEY";

// Actions
export function newWalletKeys(passphrase) {
  return {
    type: NEW_WALLET_KEYS,
    passphrase
  };
}

export function newWallet(account) {
  return {
    type: NEW_WALLET,
    wif: account.wif,
    address: account.address,
    passphrase: account.passphrase,
    encryptedWif: account.encryptedWif
  };
}

export function generating(bool) {
  return {
    type: SET_GENERATING,
    state: bool
  };
}

export function resetKey() {
  return {
    type: RESET_KEY
  };
}

// Reducer used for state necessary to generating a wallet
export default (
  state = {
    wif: null,
    address: null,
    passphrase: null,
    encryptedWif: null,
    generating: false
  },
  action
) => {
  switch (action.type) {
    case NEW_WALLET_KEYS:
      const newPrivateKey = generatePrivateKey();
      const newWif = getWIFFromPrivateKey(newPrivateKey);
      const encryptedWif = encrypt_wif(newWif, action.passphrase);
      const loadAccount = getAccountsFromWIFKey(newWif);
      return {
        ...state,
        wif: newWif,
        address: loadAccount[0].address,
        passphrase: action.passphrase,
        encryptedWif
      };
    case NEW_WALLET:
      return {
        ...state,
        wif: action.wif,
        address: action.address,
        passphrase: action.passphrase,
        encryptedWif: action.encryptedWif,
        generating: false
      };
    case SET_GENERATING:
      return { ...state, generating: action.state };
    case RESET_KEY:
      return {
        wif: null,
        address: null,
        passphrase: null,
        encryptedWif: null,
        generating: false
      };
    default:
      return state;
  }
};
