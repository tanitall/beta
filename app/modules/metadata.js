// Constants
export const SET_HEIGHT = "SET_HEIGHT";
export const SET_NETWORK = "SET_NETWORK";
export const SET_EXPLORER = "SET_EXPLORER";

// Actions
export function setNetwork(net) {
  const network = net === "MainNet" ? "MainNet" : "TestNet";
  return {
    type: SET_NETWORK,
    net: network
  };
}

export function setBlockHeight(blockHeight) {
  return {
    type: SET_HEIGHT,
    blockHeight
  };
}

export function setBlockExplorer(blockExplorer) {
  return {
    type: SET_EXPLORER,
    blockExplorer
  };
}

// reducer for metadata associated with Neon
export default (
  state = { blockHeight: 0, network: "TestNet", blockExplorer: "Neotracker" },
  action
) => {
  switch (action.type) {
    case SET_HEIGHT:
      return { ...state, blockHeight: action.blockHeight };
    case SET_EXPLORER:
      console.log(action.blockExplorer);
      return { ...state, blockExplorer: action.blockExplorer };
    case SET_NETWORK:
      return { ...state, network: action.net };
    default:
      return state;
  }
};
