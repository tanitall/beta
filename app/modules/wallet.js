// Constants
export const SET_BALANCE = "SET_BALANCE";
export const SET_MARKET_PRICE = "SET_MARKET_PRICE";
export const RESET_PRICE = "RESET_PRICE";
export const SET_TRANSACTION_HISTORY = "SET_TRANSACTION_HISTORY";

// Actions
export function setBalance(neo, gas, price, combined, gasPrice) {
  return {
    type: SET_BALANCE,
    Neo: neo,
    Gas: gas,
    price: price,
    combined: combined,
    gasPrice: gasPrice
  };
}

export function setMarketPrice(price) {
  return {
    type: SET_MARKET_PRICE,
    price: price
  };
}

export function resetPrice() {
  return {
    type: RESET_PRICE
  };
}

export function setTransactionHistory(transactions) {
  return {
    type: SET_TRANSACTION_HISTORY,
    transactions
  };
}

// reducer for wallet account balance
export default (
  state = {
    Neo: 0,
    Gas: 0,
    transactions: [],
    price: "--",
    combined: "--",
    gasPrice: "--"
  },
  action
) => {
  switch (action.type) {
    case SET_BALANCE:
      return {
        ...state,
        Neo: action.Neo,
        Gas: action.Gas,
        price: action.price,
        combined: action.combined,
        gasPrice: action.gasPrice
      };
    case RESET_PRICE:
      return { ...state, price: "--" };
    case SET_MARKET_PRICE: //current market price action type
      let currentPrice;
      if (action.price !== undefined) {
        currentPrice = action.price;
      } else {
        currentPrice = "--";
      }
      return { ...state, price: currentPrice };
    case SET_TRANSACTION_HISTORY:
      return { ...state, transactions: action.transactions };
    default:
      return state;
  }
};
