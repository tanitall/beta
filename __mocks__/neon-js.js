import { SEND_TRANSACTION } from '../app/modules/transactions';

const neonjs = jest.genMockFromModule('neon-js');

const promiseMockGen = (result, error = false) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      if (error) reject();
      resolve(result);
    });
  });
};

const privateKey = 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn';
const address = 'AM22coFfbe9N6omgL9ucFBLkeaMNg9TEyL';
const encryptedKey = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu';

neonjs.getTransactionHistory = promiseMockGen([]);
neonjs.getClaimAmounts = promiseMockGen({ available: 0, unavailable: 0 });
neonjs.getWalletDBHeight = promiseMockGen(586435);
neonjs.getBalance = promiseMockGen({ Neo: 1, Gas: 1 });
neonjs.getAPIEndpoint = jest.fn(() => 'http://testnet-api.wallet.cityofzion.io');
neonjs.decrypt_wif = jest.fn((wif) => {
  return new Promise((resolve, reject) => {
    if (!wif) reject();
    resolve(privateKey);
  });
});
neonjs.getAccountsFromWIFKey = jest.fn(() => [{ address }]);
neonjs.generatePrivateKey = jest.fn();
neonjs.getWIFFromPrivateKey = jest.fn(() => privateKey);
neonjs.encrypt_wif = jest.fn(() => encryptedKey);

module.exports = neonjs;
