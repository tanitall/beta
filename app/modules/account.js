import { getAccountsFromWIFKey } from "neon-js";

// Constants
export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";
export const SET_DECRYPTING = "SET_DECRYPTING";
export const SET_KEYS = "SET_KEYS";

// Actions
export function login(wif) {
  return {
    type: LOGIN,
    wif: wif
  };
}

export function logout() {
  return {
    type: LOGOUT
  };
}

export function decrypting(bool) {
  return {
    type: SET_DECRYPTING,
    state: bool
  };
}

export function setKeys(keys) {
  return {
    type: SET_KEYS,
    keys
  };
}

// Reducer that manages account state (account now = private key)
export default (
  state = {
    wif: null,
    address: null,
    loggedIn: false,
    redirectUrl: null,
    decrypting: false,
    accountKeys: []
  },
  action
) => {
  switch (action.type) {
    case LOGIN:
      let loadAccount;
      try {
        loadAccount = getAccountsFromWIFKey(action.wif)[0];
      } catch (e) {
        loadAccount = -1;
      }
      if (
        loadAccount === -1 ||
        loadAccount === -2 ||
        loadAccount === undefined
      ) {
        return { ...state, wif: action.wif, loggedIn: false };
      }
      return {
        ...state,
        wif: action.wif,
        address: loadAccount.address,
        loggedIn: true,
        decrypting: false
      };
    case LOGOUT:
      return {
        ...state,
        wif: null,
        address: null,
        loggedIn: false,
        decrypting: false
      };
    case SET_DECRYPTING:
      return { ...state, decrypting: action.state };
    case SET_KEYS:
      return { ...state, accountKeys: action.keys };
    default:
      return state;
  }
};
