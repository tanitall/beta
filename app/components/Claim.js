import React, { Component } from "react";
import { connect } from "react-redux";
import { setClaimRequest, disableClaim } from "../modules/claim";
import { sendEvent, clearTransactionEvent } from "../modules/transactions";
import { doClaimAllGas, doSendAsset } from "neon-js";
import ReactTooltip from "react-tooltip";
import { log } from "../util/Logs";

// wrap claiming with notifications

const doClaimNotify = (dispatch, net, selfAddress, wif) => {
  log(net, "CLAIM", selfAddress, { info: "claim all gas" });
  doClaimAllGas(net, wif).then(response => {
    if (response.result === true) {
      dispatch(
        sendEvent(
          true,
          "Your GAS claim was successful! Your balance will update once the blockchain has processed it."
        )
      );
      setTimeout(() => dispatch(disableClaim(false)), 3000);
    } else {
      dispatch(sendEvent(false, "Sorry. Claim failed. Please try again."));
    }
    setTimeout(() => dispatch(clearTransactionEvent()), 5000);
  });
};

// To initiate claim, first send all Neo to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
const doGasClaim = (dispatch, net, wif, selfAddress, ans) => {
  // if no neo in account, no need to send to self first
  if (ans === 0) {
    doClaimNotify(dispatch, net, selfAddress, wif);
  } else {
    dispatch(sendEvent(true, "Sending 0 Neo in order to claim your GAS..."));
    log(net, "SEND", selfAddress, {
      to: selfAddress,
      amount: ans,
      asset: "NEO"
    });
    doSendAsset(net, selfAddress, wif, "Neo", ans).then(response => {
      if (response.result === undefined || response.result === false) {
        dispatch(
          sendEvent(false, "Oops! Transaction failed. Please try again.")
        );
      } else {
        dispatch(sendEvent(true, "Waiting for the transaction to clear..."));
        dispatch(setClaimRequest(true));
        dispatch(disableClaim(true));
      }
    });
  }
};

class Claim extends Component {
  componentDidUpdate = () => {
    if (
      this.props.claimRequest === true &&
      this.props.claimWasUpdated == true
    ) {
      this.props.dispatch(setClaimRequest(false));
      doClaimNotify(
        this.props.dispatch,
        this.props.net,
        this.props.address,
        this.props.wif
      );
    }
  };

  render = () => {
    let renderButton;
    const doClaim = () =>
      doGasClaim(
        this.props.dispatch,
        this.props.net,
        this.props.wif,
        this.props.address,
        this.props.neo
      );
    if (this.props.disableClaimButton === false) {
      renderButton = (
        <div id="gas-button" onClick={doClaim}>
          <span className="gas-claim">
            Claim Gas<br />
            {this.props.claimAmount}
          </span>
        </div>
      );
    } else {
      renderButton = (
        <div>
          <div id="gas-button" data-tip data-for="claimTip" className="">
            <span className="gas-claim">
              Claim Gas<br />
              {this.props.claimAmount}
            </span>
          </div>
          <ReactTooltip
            className="solidTip"
            id="claimTip"
            place="bottom"
            type="dark"
            effect="solid"
          >
            <span>You can claim GAS once every 5 minutes</span>
          </ReactTooltip>
        </div>
      );
    }
    return <div id="claim">{renderButton}</div>;
  };
}

const mapStateToProps = state => ({
  claimAmount: state.claim.claimAmount,
  claimRequest: state.claim.claimRequest,
  claimWasUpdated: state.claim.claimWasUpdated,
  disableClaimButton: state.claim.disableClaimButton,
  wif: state.account.wif,
  address: state.account.address,
  net: state.metadata.network,
  neo: state.wallet.Neo
});

Claim = connect(mapStateToProps)(Claim);

export default Claim;
