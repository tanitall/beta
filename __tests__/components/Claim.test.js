import React from 'react';
import _ from 'lodash';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import Claim from '../../app/components/Claim';
import * as neonjs from 'neon-js';
import { setClaimRequest, disableClaim } from '../../app/modules/claim';
import { sendEvent } from '../../app/modules/transactions';

const state = { 
  claim: {
    claimAmount: 10,
    claimRequest: false,
    claimWasUpdated: false,
    disableClaimButton: false,
  },
  account: {
    wif: 'wif',
    address: 'address'
  },
  metadata: {
    network: 'network',
  },
  wallet: {
    Neo: 'neo',
  }
}

describe('Claim', () => {
  test('should render without crashing', () => {
    const store = configureStore()(state);
    const wrapper = shallow(<Claim store={store} />);
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should render claim gas button when claim button is not disabled', () => {
    const store = configureStore()(state);
    const wrapper = shallow(<Claim store={store} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });
  
  test('should not render claim gas button when claim button is disabled', () => {
    const newState = _.cloneDeep(state);
    newState.claim.disableClaimButton = true;
    const store = configureStore()(newState);
    const wrapper = shallow(<Claim store={store} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });
  
  describe('when do gas claim button is clicked', () => {
    const store = configureStore()(state);
    const wrapper = shallow(<Claim store={store} />);
    
    test('should dispatch transaction failure event', (done) => {
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: undefined });
        });
      });
      wrapper.dive().find('#claim button').simulate('click');
      
      setTimeout(() => {
        try {
          expect(store.getActions()[0]).toEqual(sendEvent(true, 'Sending Neo to Yourself...')); 
          expect(store.getActions()[1]).toEqual(sendEvent(false, 'Transaction failed!')); 
        } catch (error) {
          throw 'TimeoutException';
        } finally {
          done();
        }
      }, 0);
    });
    
    test('should dispatch transaction waiting, set claim request and disable claim event', (done) => {
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: '' });
        });
      });
      wrapper.dive().find('#claim button').simulate('click');

      setTimeout(() => {
        try {
          expect(store.getActions()[0]).toEqual(sendEvent(true, 'Sending Neo to Yourself...')); 
          expect(store.getActions()[1]).toEqual(sendEvent(true, "Waiting for transaction to clear...")); 
          expect(store.getActions()[2]).toEqual(setClaimRequest(true)); 
          expect(store.getActions()[3]).toEqual(disableClaim(true)); 
        } catch (error) {
          throw 'TimeoutException';
        } finally {
          done();
        }
      }, 0);
    });
  });
  
  describe('when claim is requested and updated', () => {
    const newState = _.cloneDeep(state);
    newState.claim.claimRequest = true;
    newState.claim.claimWasUpdated = true;
    newState.dispatch = jest.fn();
    
    test('should dispatch false claim request and claim successful event', (done) => {
      const store = configureStore()(newState);
      const wrapper = mount(<Claim store={store} />);
      neonjs.doClaimAllGas = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: true });
        });
      });
      wrapper.setProps({ target: '' });
      
      setTimeout(() => {
        try {
          expect(store.getActions()[0]).toEqual(setClaimRequest(false)); 
          expect(store.getActions()[1]).toEqual(sendEvent(true, 'Claim was successful! Your balance will update once the blockchain has processed it.')); 
        } catch (error) {
          throw 'TimeoutException';
        } finally {
          done();
        }
      }, 0);
    });
    
    test('should dispatch false claim request and claim failure event', (done) => {
      const store = configureStore()(newState);
      const wrapper = mount(<Claim store={store} />);
      neonjs.doClaimAllGas = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: false });
        });
      });
      wrapper.setProps({ target: '' });
      
      setTimeout(() => {
        try {
          expect(store.getActions()[0]).toEqual(setClaimRequest(false)); 
          expect(store.getActions()[1]).toEqual(sendEvent(false, 'Claim failed')); 
        } catch (error) {
          throw 'TimeoutException';
        } finally {
          done();
        }
      }, 0);
    });
  });
});