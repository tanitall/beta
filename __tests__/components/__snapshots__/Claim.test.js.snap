// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim should not render claim gas button when claim button is disabled 1`] = `
<div
  id="claim"
>
  <div>
    <button
      className="disabled"
      data-for="claimTip"
      data-tip={true}
    >
      Claim 
      10
       GAS
    </button>
    <ReactTooltip
      class="solidTip"
      effect="solid"
      id="claimTip"
      insecure={true}
      place="bottom"
      resizeHide={true}
      type="dark"
      wrapper="div"
    >
      <span>
        You can claim Gas once every 5 minutes
      </span>
    </ReactTooltip>
  </div>
</div>
`;

exports[`Claim should render claim gas button when claim button is not disabled 1`] = `
<div
  id="claim"
>
  <button
    onClick={[Function]}
  >
    Claim 
    10
     GAS
  </button>
</div>
`;

exports[`Claim should render without crashing 1`] = `
<Claim
  address="address"
  claimAmount={10}
  claimRequest={false}
  claimWasUpdated={false}
  disableClaimButton={false}
  dispatch={[Function]}
  neo="neo"
  net="network"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  wif="wif"
/>
`;
